scene/resources/material.cpp

```diff
@@ -761,11 +761,26 @@ void SpatialMaterial::_update_shader() {
 			code += "\t\tvec2 ofs = base_uv;\n";
 			code += "\t\tfloat depth = textureLod(texture_depth, ofs, 0.0).r;\n";
 			code += "\t\tfloat current_depth = 0.0;\n";
-			code += "\t\twhile(current_depth < depth) {\n";
-			code += "\t\t\tofs -= delta;\n";
-			code += "\t\t\tdepth = textureLod(texture_depth, ofs, 0.0).r;\n";
-			code += "\t\t\tcurrent_depth += layer_depth;\n";
-			code += "\t\t}\n";
+			if (VisualServer::get_singleton()->is_low_end()) {
+				// `while` loops are not well supported in GLES2. Do a "large" finite
+				// loop with `for`, but not too large, as it slows down some compilers.
+				// The loop's size is tailored to match the default number of maximum steps (32).
```

Comment:
We could set the loop to run 64 iterations at most (so that we don't need to limit the number of deep parallax layers in GLES2), but this may slow down shader compilation times on some devices. Therefore, I've decided to limit to 32. The renderer is low-end oriented anyway, so values higher than 32 are unlikely to be usable in real world scenarios.